# Statistical Methods in Pandas - Lab

## Introduction

In this lab you'll get some hands-on experience using some of the key summary statistics methods in Pandas.

## Objectives
You will be able to:

- Calculate summary statistics for a series and DataFrame 
- Use the `.apply()` or `.applymap()` methods to apply a function to a pandas series or DataFrame  


## Getting Started

For this lab, we'll be working with a dataset containing information on various lego datasets. You will find this dataset in the file `'lego_sets.csv'`.   

In the cell below:

- Import Pandas and set the standard alias of `pd`
- Import the `'lego_sets.csv'` dataset 
- Display the first five rows of the DataFrame to get a feel for what we'll be working with

In [12]:
# Import pandas
import pandas as pd

# Import the 'lego_sets.csv' dataset
df = pd.read_csv('lego_sets.csv')

# Print the first five rows of DataFrame
df.head()
df.tail()

ages  list_price  num_reviews  piece_count  play_star_rating  \
12256  7-14     36.5878          6.0        341.0               4.4   
12257  7-14     24.3878          8.0        217.0               4.1   
12258  7-14     24.3878         18.0        233.0               4.6   
12259  6-14     12.1878          1.0         48.0               5.0   
12260  6-14     12.1878         11.0        109.0               4.5   

                                               prod_desc  prod_id  \
12256  Protect NINJAGO® City from flying Manta Ray Bo...  70609.0   
12257         Stop a Piranha Attack with Kai and Misako!  70629.0   
12258   Stop a crime in the NINJAGO® City street market!  70607.0   
12259  Achieve Spinjitzu greatness with the Green Ninja!  70628.0   
12260  Practice your Spinjitzu skills with Kai and Zane!  70606.0   

                                          prod_long_desc review_difficulty  \
12256  Help Cole save Shen-Li in this cool THE LEGO® ...              Easy   
12257  Play out an action-packed Piranha Mech pursuit...              Easy   
12258  Team up with Lloyd Garmadon, Nya and Officer T...              Easy   
12259  Learn all the skills of Spinjitzu with THE LEG...         Very Easy   
12260  Join the ninja heroes at the dojo with this ac...              Easy   

                       set_name  star_rating                 theme_name  \
12256          Manta Ray Bomber          4.3  THE LEGO® NINJAGO® MOVIE™   
12257            Piranha Attack          3.6  THE LEGO® NINJAGO® MOVIE™   
12258       NINJAGO® City Chase          4.6  THE LEGO® NINJAGO® MOVIE™   
12259  Lloyd - Spinjitzu Master          5.0  THE LEGO® NINJAGO® MOVIE™   
12260        Spinjitzu Training          4.7  THE LEGO® NINJAGO® MOVIE™   

       val_star_rating country  
12256              4.2      PT  
12257              4.1      PT  
12258              4.5      PT  
12259              5.0      PT  
12260              4.8      PT

## Getting DataFrame-Level Statistics

We'll begin by getting some overall summary statistics on the dataset. There are two ways we'll get this information -- `.info()` and `.describe()`.

The `.info()` method provides us metadata on the DataFrame itself. This allows us to answer questions such as:

* What data type does each column contain?
* How many rows are in my dataset? 
* How many total non-missing values does each column contain?
* How much memory does the DataFrame take up?

In the cell below, call our DataFrame's `.info()` method. 

In [2]:
# Call the .info() method
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ages               12261 non-null  object 
 1   list_price         12261 non-null  float64
 2   num_reviews        10641 non-null  float64
 3   piece_count        12261 non-null  float64
 4   play_star_rating   10486 non-null  float64
 5   prod_desc          11884 non-null  object 
 6   prod_id            12261 non-null  float64
 7   prod_long_desc     12261 non-null  object 
 8   review_difficulty  10206 non-null  object 
 9   set_name           12261 non-null  object 
 10  star_rating        10641 non-null  float64
 11  theme_name         12258 non-null  object 
 12  val_star_rating    10466 non-null  float64
 13  country            12261 non-null  object 
dtypes: float64(7), object(7)
memory usage: 1.3+ MB


#### Interpreting the Results

Read the output above, and then answer the following questions:

- How many total rows are in this DataFrame?
    - 12261 total entries
- How many columns contain numeric data?
    - 7
- How many contain categorical data? 
    - 7
- Identify at least 3 columns that contain missing values.
    - num_reviews, play_star_rating, prod_desc



Whereas `.info()` provides statistics about the DataFrame itself, `.describe()` returns output containing basic summary statistics about the data contained with the DataFrame.  

In the cell below, call the DataFrame's `.describe()` method. 

In [3]:
# Call the .describe() method
df.describe()


list_price   num_reviews   piece_count  play_star_rating  \
count  12261.000000  10641.000000  12261.000000      10486.000000   
mean      65.141998     16.826238    493.405921          4.337641   
std       91.980429     36.368984    825.364580          0.652051   
min        2.272400      1.000000      1.000000          1.000000   
25%       19.990000      2.000000     97.000000          4.000000   
50%       36.587800      6.000000    216.000000          4.500000   
75%       70.192200     13.000000    544.000000          4.800000   
max     1104.870000    367.000000   7541.000000          5.000000   

            prod_id   star_rating  val_star_rating  
count  1.226100e+04  10641.000000     10466.000000  
mean   5.983675e+04      4.514134         4.228960  
std    1.638115e+05      0.518865         0.660282  
min    6.300000e+02      1.800000         1.000000  
25%    2.103400e+04      4.300000         4.000000  
50%    4.206900e+04      4.700000         4.300000  
75%    7.092200e+04      5.000000         4.700000  
max    2.000431e+06      5.000000         5.000000

#### Interpreting the Results

The output contains descriptive statistics corresponding to the columns. Use these to answer the following questions:

- How much is the standard deviation for `piece count`? 
    - 825.4
- How many pieces are in the largest lego set?
    - 7541
- How many in the smallest lego set? What is the median `val_star_rating`?
    - 1
    - 4.2 

________________________________________________________________________________________________________________________________

## Getting Summary Statistics

Pandas also allows us to easily compute individual summary statistics using built-in methods.  Next, we'll get some practice using these methods. 

In the cell below, compute the median value of the `star_rating` column.

In [4]:
# Calculate the median of the star_rating column
df['star_rating'].mean()

4.514134009961469

Next, get a count of the total number of unique values in `play_star_rating`.

In [9]:
# Print the number of unique values in play_star_rating
len(df['play_star_rating'].unique())

31

Now, compute the standard deviation of the `list_price` column.

In [10]:
# Calculate the standard deviation of the list_price column
df['list_price'].std()

91.9804293059243

If we bought every single lego set in this dataset, how many pieces would we have?  

> **Note**: If you truly want to answer this accurately, and are up for the challenge, remove duplicate lego-set entries before summing the pieces. That is, many of the lego sets are listed multiple times in the dataset above, depending on the country where it is being sold and other unique parameters. If you're stuck, just practice calculating the total number of pieces in the dataset for now.

In [11]:
# Total number of pieces across all unique Lego sets
df['piece_count'].sum()

6049650.0

In [15]:
len(df['prod_id'].unique())

744

Now, let's try getting the value for the 90% quantile for all numerical columns.  Do this in the cell below.

In [19]:
# Get the 90% quantile for all numerical columns

numer_cols = df.select_dtypes(include=['float64', 'int64'])
numer_cols.quantile(0.9)

list_price            136.2971
num_reviews            38.0000
piece_count          1077.0000
play_star_rating        5.0000
prod_id             75531.0000
star_rating             5.0000
val_star_rating         5.0000
Name: 0.9, dtype: float64

## Getting Summary Statistics on Categorical Data

For obvious reasons, most of the methods we've used so far only work with numerical data -- there's no way to calculate the standard deviation of a column containing string values. However, there are some things that we can discover about columns containing categorical data. 

In the cell below, print the unique values contained within the `review_difficulty` column. 

In [20]:
# Print the unique values in the review_difficulty column
df['review_difficulty'].unique()

array(['Average', 'Easy', 'Challenging', 'Very Easy', nan,
       'Very Challenging'], dtype=object)

Now, let's get the `value_counts()` for this column, to see how common each is. 

In [21]:
# Get the value_counts() of the review_difficulty column
df['review_difficulty'].value_counts()

Easy                4236
Average             3765
Very Easy           1139
Challenging         1058
Very Challenging       8
Name: review_difficulty, dtype: int64

As you can see, these provide us quick and easy ways to get information on columns containing categorical information.  


## Using `.applymap()`

When working with pandas DataFrames, we can quickly compute functions on the data contained by using the `.applymap()` method and passing in a lambda function. 

For instance, we can use `applymap()` to return a version of the DataFrame where every value has been converted to a string.

In the cell below:

* Call the DataFrame's `.applymap()` method and pass in `lambda x: str(x)`  
* Call the new `string_df` object's `.info()` method to confirm that everything has been cast to a string

In [22]:
# Call the .applymap() method
string_df = df.applymap(lambda x: str(x))

# Call the .info() method
string_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   ages               12261 non-null  object
 1   list_price         12261 non-null  object
 2   num_reviews        12261 non-null  object
 3   piece_count        12261 non-null  object
 4   play_star_rating   12261 non-null  object
 5   prod_desc          12261 non-null  object
 6   prod_id            12261 non-null  object
 7   prod_long_desc     12261 non-null  object
 8   review_difficulty  12261 non-null  object
 9   set_name           12261 non-null  object
 10  star_rating        12261 non-null  object
 11  theme_name         12261 non-null  object
 12  val_star_rating    12261 non-null  object
 13  country            12261 non-null  object
dtypes: object(14)
memory usage: 1.3+ MB


Note that everything -- even the `NaN` values, have been cast to a string in the example above. 

Note that for Pandas Series objects (such as a single column in a DataFrame), we can do the same thing using the `.apply()` method.  

This is just one example of how we can quickly compute custom functions on our DataFrame -- this will become especially useful when we learn how to **_normalize_** our datasets in a later section!

## Summary

In this lab, we learned how to:

* Use the `df.describe()` and `df.info()` summary statistics methods 
* Use built-in Pandas methods for calculating summary statistics 
* Apply a function to every element in a DataFrame